File: src/javascript/algorithms/linkedlist/findKthNodeFromEnd.js
class Node {
    constructor(value) {
        this.value = value;
        this.next = null;
    }
}

class LinkedList {
    constructor(value) {
        this.head = null;
        this.tail = null;
        this.length = 0;
    }

    push(value) {
        const newNode = new Node(value);
        if(!this.head) {
            this.head = newNode;
            this.tail = newNode;
        } else {
            this.tail.next = newNode;
            this.tail = newNode;
        }
        this.length++;
        return this;
    }

    findKthNodeFromEnd(k) {
        let slow = this.head;
        let fast = this.head;
        for (let i = 0; i < k; i++) {
            if (fast === null) return null;
            fast = fast.next;
        }

        while(fast != null) {
            slow = slow.next;
            fast = fast.next;
        }
        return slow;
    }
}

const myLinkedList = new LinkedList();
myLinkedList.push(1);
myLinkedList.push(2);
myLinkedList.push(3);
myLinkedList.push(4);
myLinkedList.push(5);

myLinkedList.findKthNodeFromEnd(3);

File: src/java1/algorithms/linkedlist/FindKthNodeFromEnd.java
package java1.algorithms.linkedlist;

public class FindKthNodeFromEnd {
    private static Node findKthNodeFromEnd(Node head, int k) {
        Node slowPtr = head;
        Node fastPtr = head;

        for(int i=0; i<k; i++){
            if(fastPtr == null) return null;
            fastPtr = fastPtr.next;
        }

        while (fastPtr != null) {
            slowPtr = slowPtr.next;
            fastPtr = fastPtr.next;
        }
        return slowPtr;
    }

    public static void main(String[] args) {
        Node headNode = new Node(10);
        Node node2 = new Node(20);
        Node node3 = new Node(30);
        Node node4 = new Node(40);
        Node node5 = new Node(50);
        Node node6 = new Node(60);
        Node node7 = new Node(70);
        Node node8 = new Node(80);

        headNode.setNext(node2);
        node2.setNext(node3);
        node3.setNext(node4);
        node4.setNext(node5);
        node5.setNext(node6);
        node6.setNext(node7);
        node7.setNext(node8);

        System.out.println(findKthNodeFromEnd(headNode, 3).value);
    }
}
